### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by ducphamle2**

### CODE REVIEW

**Review by marxeille — APPROVED** [review]

File: packages/cosmos/src/tx-tracer/index.ts [hunk 1]
```diff
@@ -51,16 +52,16 @@ export class TendermintTxTracer {
   }
 
   protected getWsEndpoint(): string {
-    let url = this.url;
-    if (url.startsWith('http')) {
+    let url = this.url ?? 'https://rpc.orai.io';
```
Comment (tubackkhoa) [inline packages/cosmos/src/tx-tracer/index.ts:55]: why default url is 'https://rpc.orai.io', must only subscribe when url is not empty

File: packages/cosmos/src/tx-tracer/index.ts [hunk 2]
```diff
@@ -51,16 +52,16 @@ export class TendermintTxTracer {
...
       url = url.replace('http', 'ws');
     }
-    if (!url.endsWith(this.wsEndpoint)) {
+    if (!url?.endsWith(this.wsEndpoint)) {
       const wsEndpoint = this.wsEndpoint.startsWith('/')
         ? this.wsEndpoint
         : '/' + this.wsEndpoint;
 
-      url = url.endsWith('/') ? url + wsEndpoint.slice(1) : url + wsEndpoint;
+      url = url?.endsWith('/') ? url + wsEndpoint.slice(1) : url + wsEndpoint;
```
Comment (tubackkhoa) [inline packages/cosmos/src/tx-tracer/index.ts:64]: should trim '/' and join array[string] with '/', arr.map(item=>item.replace(/\/$/,'')).join('/')

File: packages/cosmos/src/tx-tracer/index.ts [hunk 3]
```diff
@@ -105,7 +106,11 @@ export class TendermintTxTracer {
     }
 
     for (const [id, tx] of this.txSubscribes) {
-      this.sendSubscribeTxRpc(id, tx.hash);
+      if (tx.hash) {
+        this.sendSubscribeTxRpc(id, tx.hash);
+      } else {
+        this.sendSubscribeMsgRpc(id, tx.address);
```
Comment (tubackkhoa) [inline packages/cosmos/src/tx-tracer/index.ts:112]: address is optional, why not check tx.address ?

File: packages/cosmos/src/tx-tracer/index.ts [hunk 4]
```diff
@@ -212,9 +217,51 @@ export class TendermintTxTracer {
     }
   }
 
+  //  Subscribe the msg
+  subscribeMsgByAddress(msg: string): Promise<any> {
+    return new Promise<any>(resolve => {
+      this.subscribeMsg(msg).then(resolve);
```
Comment (tubackkhoa) [inline packages/cosmos/src/tx-tracer/index.ts:223]: not return promise, how come then afterward?, and reject should be added.
Better using this: await Promise.race([timeout, this.subscribeMsg(msg)])

File: packages/extension/src/pages/main/token.tsx [hunk 5]
```diff
@@ -241,8 +241,20 @@ export const TokensView: FunctionComponent<{
...
+      const denomHelper = new DenomHelper(
+        v.balance.trim(true).shrink(true).currency.coinMinimalDenom
+      );
+      return (
+        (denomHelper.contractAddress
+          ? obj.findIndex(
+              (v2) =>
+                new DenomHelper(
+                  v2.balance.trim(true).shrink(true).currency.coinMinimalDenom
+                ).contractAddress === denomHelper.contractAddress
```
Comment (tubackkhoa) [inline packages/extension/src/pages/main/token.tsx:254]: why not compare coinMinimalDenom directly, what happend if coinMinimalDenom is 'uorai' so contractAddress is empty and still duplicate

File: packages/extension/src/pages/register/advanced-bip44.tsx [hunk 6]
```diff
@@ -8,7 +8,7 @@ import { BIP44HDPath } from '@owallet/background';
 
 export class BIP44Option {
   @observable
-  protected _coinType?: number;
+  protected _coinType?: number = 0;
```
Comment (tubackkhoa) [inline packages/extension/src/pages/register/advanced-bip44.tsx:11]: if there is default value, type is not optional anymore, so it must be protected _coinType: number = 0;

File: packages/extension/src/pages/register/advanced-bip44.tsx [hunk 7]
```diff
@@ -144,13 +148,49 @@ export const AdvancedBIP44Option: FunctionComponent<{
...
+                  ? bip44Option?.coinType?.toString()
+                  : 0
+              }
+              onChange={(e) => {
+                e.preventDefault();
+
+                let value = e.target.value;
+                if (value) {
+                  if (value !== '0') {
+                    // Remove leading zeros
```
Comment (tubackkhoa) [inline packages/extension/src/pages/register/advanced-bip44.tsx:170]: better using if (value.replace(/^0+/,'')) { }

File: packages/extension/src/pages/send/index.tsx [hunk 8]
```diff
@@ -216,6 +216,18 @@ export const SendPage: FunctionComponent<{
                       tx,
                       'TX INFO ON SEND PAGE!!!!!!!!!!!!!!!!!!!!!'
                     );
+                    notification.push({
+                      placement: 'top-center',
+                      type: tx?.data ? 'success' : 'danger',
+                      duration: 5,
+                      content: tx?.data
+                        ? `Transaction successful with tx: ${tx?.hash}`
```
Comment (tubackkhoa) [inline packages/extension/src/pages/send/index.tsx:224]: why not check tx.code === 0 ? success : error

File: packages/mobile/src/screens/register/mnemonic/recover-mnemonic.tsx [hunk 9]
```diff
@@ -24,7 +24,7 @@ import { LoadingSpinner } from '../../../components/spinner';
 const bip39 = require('bip39');
 
 function isPrivateKey(str: string): boolean {
-  if (str.startsWith('0x')) {
+  if (str?.startsWith('0x')) {
```
Comment (tubackkhoa) [inline packages/mobile/src/screens/register/mnemonic/recover-mnemonic.tsx:27]: this is wrong, must use export const isPrivateKey = (str: string): boolean =>
  /(?:0x)?[0-9a-fA-F]{64}/.test(str); package/common/src/utils/strings.ts

File: packages/mobile/src/screens/setting/screens/select-account/index.tsx [hunk 10]
```diff
@@ -84,19 +85,19 @@ export const SettingSelectAccountScreen: FunctionComponent = observer(() => {
...
     return keyRingStore.multiKeyStoreInfo.filter(
-      (keyStore) => keyStore.type === 'ledger'
+      keyStore => keyStore.type === 'ledger'
     );
   }, [keyRingStore.multiKeyStoreInfo]);
 
   const privateKeyStores = useMemo(() => {
     return keyRingStore.multiKeyStoreInfo.filter(
-      (keyStore) => keyStore.type === 'privateKey' && !keyStore.meta?.email
+      keyStore => keyStore.type === 'privateKey' && !keyStore.meta?.email
```
Comment (tubackkhoa) [inline packages/mobile/src/screens/setting/screens/select-account/index.tsx:100]: remove keyStore.meta?.email, keyStore.type social is removed, no meta.email anymore.

File: packages/mobile/src/screens/transactions/components/items/items.tsx [hunk 11]
```diff
@@ -17,45 +25,144 @@ interface TransactionItemProps {
 export const TransactionItem: FunctionComponent<TransactionItemProps> = ({
   item,
   address,
+  type,
   onPress,
   containerStyle
 }) => {
   const { txhash, tx, timestamp } = item || {};
+
   const date = moment(timestamp).format('MMM DD, YYYY [at] HH:mm');
```
Comment (tubackkhoa) [inline packages/mobile/src/screens/transactions/components/items/items.tsx:34]: better using config from languages in common package

File: packages/mobile/src/screens/transactions/details.tsx [hunk 12]
```diff
@@ -234,72 +234,213 @@ export const TransactionDetail: FunctionComponent<any> = () => {
...
+
+  const amountDataCell = useCallback(() => {
+    let amount;
+
+    if (
+      item?.messages?.find(
+        msg => getTxTypeNew(msg['@type']) === 'MsgRecvPacket'
+      )
+    ) {
+      const msg = item?.messages?.find(msg => {
```
Comment (tubackkhoa) [inline packages/mobile/src/screens/transactions/details.tsx:258]: duplicate code, why not `if (msg = item?.messages?.find(
        msg => getTxTypeNew(msg['@type']) === 'MsgRecvPacket'
))`

**Review by tubackkhoa — COMMENTED** [review]
Need to fix resolve problem and comparing contract address for filtering token balances.

File: packages/extension/src/pages/send/index.tsx (hunk 8 above)
Comment (haunv3) [reply packages/extension/src/pages/send/index.tsx:224]: Screen Shot 2022-09-26 at 17 05 24


> 
because the data has no code

**Review by haunv3 — DISMISSED** [review]

**Review by marxeille — APPROVED** [review]

**Review by haunv3 — APPROVED** [review]

**Review by tubackkhoa — APPROVED** [review]

**Pull request merged**